File: Games/Sevens/src/main/java/jokatu/game/games/sevens/player/SevensPlayer.java
package jokatu.game.games.sevens.player;

import jokatu.game.cards.Card;
import jokatu.game.event.GameEvent;
import jokatu.game.games.sevens.event.CardDrawnEvent;
import jokatu.game.games.sevens.event.CardPlayedEvent;
import jokatu.game.player.Player;
import ophelia.event.observable.AbstractSynchronousObservable;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class SevensPlayer extends AbstractSynchronousObservable<GameEvent> implements Player {

	private final String name;
	private final List<Card> hand = new ArrayList<>();

	public SevensPlayer(String username) {
		this.name = username;
	}

	public void drawCard(Card card) {
		hand.add(card);
		fireEvent(new CardDrawnEvent(this, card));
	}

	@NotNull
	@Override
	public String getName() {
		return name;
	}

	public List<Card> getHand() {
		return hand;
	}

	public void playCard(@NotNull Card card) {
		boolean removed = hand.remove(card);
		if (removed) {
			fireEvent(new CardPlayedEvent(this, card));
		}
	}

	@Override
	public String toString() {
		return name;
	}
}

File: Games/Sevens/src/main/kotlin/jokatu/game/games/sevens/player/SevensPlayer.kt
package jokatu.game.games.sevens.player

import jokatu.game.cards.Card
import jokatu.game.event.GameEvent
import jokatu.game.games.sevens.event.CardDrawnEvent
import jokatu.game.games.sevens.event.CardPlayedEvent
import jokatu.game.player.Player
import ophelia.event.observable.AbstractSynchronousObservable
import java.util.*

class SevensPlayer(override val name: String) : AbstractSynchronousObservable<GameEvent>(), Player {
	val hand: MutableList<Card> = ArrayList()

	fun drawCard(card: Card) {
		hand.add(card)
		fireEvent(CardDrawnEvent(this, card))
	}

	fun playCard(card: Card) {
		val removed = hand.remove(card)
		if (removed) {
			fireEvent(CardPlayedEvent(this, card))
		}
	}

	override fun toString(): String {
		return name
	}
}
